Parallel replicas: announcement response handling improvement

devcrafter opened the pull request:
More efficient search for the same or covered/covering parts while building part's set (handling announcement response)

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

devcrafter committed:
Parallel replicas (perf): announcement response handling improvement

robot-ch-test-poll1 commented:
<!-- automatic status comment for PR #57749 from ClickHouse/ClickHouse:pr-perf-improve-hope -->
*This is an automated comment for commit 4ccf4e11a8ee0e1a7504701f0c77792064bcb5a1 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57749/4ccf4e11a8ee0e1a7504701f0c77792064bcb5a1/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7167999177">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57749/4ccf4e11a8ee0e1a7504701f0c77792064bcb5a1/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

devcrafter committed:
One lookup for 2 checks

(1) if it's the same part
(2) if annouced part has intersections with parts in working set

devcrafter commented:
@serxa sorry, I came up with a better solution which covers the case when announced part can be covering/covered /by a part from parallel replicas working set

serxa reviewed (dismissed):
This looks much better now!

devcrafter commented on src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp:
```diff
@@ -161,26 +161,37 @@ void DefaultCoordinator::updateReadingState(InitialAllRangesAnnouncement announc
+            {
+                /// check if it is covering or covered part
+                /// need to compare with 2 nearest parts in set, - lesser and greater than the part from the announcement
+                bool is_disjoint = found_part.isDisjoint(announced_part);
+                if (it != all_parts_to_read.cbegin() && is_disjoint)
+                {
+                    const MergeTreePartInfo & lesser_part_info = (--it)->description.info;
+                    is_disjoint &= lesser_part_info.isDisjoint(announced_part);
+                }
+                if (!is_disjoint)
+                    continue;
```
It looks less readable to me. And yes, the locals here are mostly for readability

devcrafter committed:
Test and fix

devcrafter committed:
Merge remote-tracking branch 'origin/master' into pr-perf-improve-hope

devcrafter committed:
Test cleanup

robot-clickhouse committed:
Automatic style fix

devcrafter committed:
Fix: stop merges, otherwise test can be flaky

devcrafter commented:
Integration test failure - https://github.com/ClickHouse/ClickHouse/issues/54465

devcrafter committed:
Comment to about possible query results

Algunenano commented on tests/integration/test_parallel_replicas_working_set/test.py:
```diff
@@ -0,0 +1,140 @@
+    expected_active_parts = ""
+    if node_with_covering_part == 0:
+        expected_active_parts = (
+            "all_0_5_1\nall_2_2_0\nall_3_3_0\nall_4_4_0\nall_5_5_0\n"
```
This will fail with ZK retries (different block number), right? You might want to disable fault injections if that's the case

devcrafter replied:
Thanks, but are we gonna to enable zk fault injections in integration test?

Algunenano replied:
True. I thought they were already on by default but they aren't.

devcrafter commented:
Test failures:
- integration test - https://github.com/ClickHouse/ClickHouse/issues/57854
